File: src/api/helpers/news-question-local-mutate.ts
import { NewsQuestion } from "~/models/news-question";
import { NewsQuestionKind } from "~/models/news-question-kind";

export const newsQuestionLocalMutate: {
  (question: NewsQuestion, textInputAnswer?: string): void;
  (question: NewsQuestion, selectedAnswers?: number[], otherFieldTextValue?: string): void;
  (question: NewsQuestion, answers?: number[] | string, textInput?: string): void;
} = (
  question: NewsQuestion,
  answers?: number[] | string,
  textInput?: string
): void => {
  if (
    question.kind === NewsQuestionKind.TextInput ||
    question.kind === NewsQuestionKind.InformationText
  ) {
    question.textInputAnswer = answers as string | undefined;
  }
  else {
    question.selectedAnswers = answers as number[] | undefined;
    question.textInputAnswer = textInput;
  }

  question.answered = typeof answers !== "undefined";
  question.answerDate = question.answered ? new Date() : undefined;
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/helpers/news-question-local-mutate.kt
package ink.literate.pawnote.api.helpers

import ink.literate.pawnote.models.NewsQuestion
import ink.literate.pawnote.models.NewsQuestionKind
import ink.literate.pawnote.models.errors.UnreachableError
import kotlinx.datetime.*

fun newsQuestionLocalMutate(
    question: NewsQuestion,
    answers: Any? = null,
    textInput: String? = null
) {
  when (question.kind) {
    NewsQuestionKind.TextInput,
    NewsQuestionKind.InformationText -> {
      question.textInputAnswer = answers as? String
    }
    else -> {
      if (answers !is List<*>) throw UnreachableError("newsQuestionLocalMutate")
      else {
        question.selectedAnswers = answers.filterIsInstance<Int>()
        question.textInputAnswer = textInput
      }
    }
  }

  question.answered = answers != null
  question.answerDate =
      if (question.answered) Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault())
      else null
}
